sdk/core/azure-core/inc/azure/core/http/curl/curl.hpp

```diff
@@ -55,405 +54,484 @@ namespace Azure { namespace Core { namespace Http {
       CURL* GetHandle() { return this->m_handle; }
 
       std::string GetHost() const { return this->m_host; }
-    };
 
-    // TODO: Mutex for this code to access connectionPoolIndex
-    /*
-     * Keeps an unique key for each host and creates a connection pool for each key.
-     * This way getting a connection for an specific host can be done in O(1) instead of looping a
-     * single connection list to find the first connection for the required host.
-     *
-     * There might be multiple connections for each host.
-     */
-    static std::map<std::string, std::list<std::unique_ptr<CurlConnection>>> s_connectionPoolIndex;
-    static std::unique_ptr<CurlConnection> GetCurlConnection(Request& request);
-    static void MoveConnectionBackToPool(std::unique_ptr<CurlSession::CurlConnection> connection);
-    static std::mutex s_connectionPoolMutex;
+      void updateLastUsageTime() { this->m_lastUsage = std::chrono::steady_clock::now(); }
 
-  private:
-    /**
-     * @brief Enum used by ResponseBufferParser to control the parsing internal state while building
-     * the HTTP RawResponse
-     *
-     */
-    enum class ResponseParserState
+      bool isExpired()
+      {
+        auto connectionOnWaitingTimeMs = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::steady_clock::now() - this->m_lastUsage);
+        if (connectionOnWaitingTimeMs.count() >= Details::c_DefaultConnectionExpiredMilliseconds)
+        {
+          return true;
+        }
+        return false;
+      }
+    };
+
+    struct CurlConnectionPool
     {
-      StatusLine,
-      Headers,
-      EndOfHeaders,
+      /*
+       * Mutex for accesing connection pool for thread-safe reading and writing
+       */
+      static std::mutex s_connectionPoolMutex;
+
+      /*
+       * Keeps an unique key for each host and creates a connection pool for each key.
+       * This way getting a connection for an specific host can be done in O(1) instead of looping a
+       * single connection list to find the first connection for the required host.
+       *
+       * There might be multiple connections for each host.
+       */
+      static std::map<std::string, std::list<std::unique_ptr<CurlConnection>>>
+          s_connectionPoolIndex;
+
+      /*
+       * Finds a connection to be re-used from the connection pool. If there is not any available
+       * connection, a new connection is created.
+       */
+      static std::unique_ptr<CurlConnection> GetCurlConnection(Request& request);
+
+      /**
+       * Moves a connection back to the pool to be re-used
+       */
+      static void MoveConnectionBackToPool(std::unique_ptr<CurlConnection> connection);
+
+#ifdef TESTING_BUILD
```

Comment:
Remove this function
This can alter behavior in the shipping code vs testing code.  Knowing the cleaner is running isn't important.  Knowing that the item gets cleaned up is all that matters.